lib/promscrape/scrapework_timing_test.go

```diff
@@ -96,3 +98,93 @@ vm_tcplistener_write_calls_total{name="https", addr=":443"} 132356
 		tsmGlobal.Unregister(&sw)
 	})
 }
+
+func BenchmarkScrapeWorkScrapeInternalStream(b *testing.B) {
+	data := `
+vm_tcplistener_accepts_total{name="http", addr=":80"} 1443
+vm_tcplistener_accepts_total{name="https", addr=":443"} 12801
+vm_tcplistener_conns{name="http", addr=":80"} 0
+vm_tcplistener_conns{name="https", addr=":443"} 2
+vm_tcplistener_errors_total{name="http", addr=":80", type="accept"} 0
+vm_tcplistener_errors_total{name="http", addr=":80", type="close"} 0
+vm_tcplistener_errors_total{name="http", addr=":80", type="read"} 97
+vm_tcplistener_errors_total{name="http", addr=":80", type="write"} 2
+vm_tcplistener_errors_total{name="https", addr=":443", type="accept"} 0
+vm_tcplistener_errors_total{name="https", addr=":443", type="close"} 0
+vm_tcplistener_errors_total{name="https", addr=":443", type="read"} 243
+vm_tcplistener_errors_total{name="https", addr=":443", type="write"} 285
+vm_tcplistener_read_bytes_total{name="http", addr=":80"} 879339
+vm_tcplistener_read_bytes_total{name="https", addr=":443"} 19453340
+vm_tcplistener_read_calls_total{name="http", addr=":80"} 7780
+vm_tcplistener_read_calls_total{name="https", addr=":443"} 70323
+vm_tcplistener_read_timeouts_total{name="http", addr=":80"} 673
+vm_tcplistener_read_timeouts_total{name="https", addr=":443"} 12353
+vm_tcplistener_write_calls_total{name="http", addr=":80"} 3996
+vm_tcplistener_write_calls_total{name="https", addr=":443"} 132356
+`
+	common.StartUnmarshalWorkers()
+	defer common.StopUnmarshalWorkers()
+
+	readDataFunc := func(dst *bytesutil.ByteBuffer) error {
+		dst.B = append(dst.B, data...)
+		return nil
+	}
+	b.ReportAllocs()
+	b.SetBytes(int64(len(data)))
+	b.RunParallel(func(pb *testing.PB) {
+		var sw scrapeWork
+		sw.Config = &ScrapeWork{
+			StreamParse: true,
+		}
+		sw.ReadData = readDataFunc
+		sw.PushData = func(_ *auth.Token, _ *prompbmarshal.WriteRequest) {}
+		tsmGlobal.Register(&sw)
+		timestamp := int64(0)
+		for pb.Next() {
+			if err := sw.scrapeInternal(timestamp, timestamp); err != nil {
+				panic(fmt.Errorf("unexpected error: %w", err))
+			}
+			timestamp++
+		}
+		tsmGlobal.Unregister(&sw)
+	})
+}
+
+func BenchmarkScrapeWorkScrapeInternalStreamBigData(b *testing.B) {
+	generateScrape := func(n int) string {
+		w := strings.Builder{}
+		for i := 0; i < n; i++ {
+			w.WriteString(fmt.Sprintf("fooooo_%d 1\n", i))
+		}
+		return w.String()
+	}
+
+	data := generateScrape(200000)
+	common.StartUnmarshalWorkers()
+	defer common.StopUnmarshalWorkers()
+
+	readDataFunc := func(dst *bytesutil.ByteBuffer) error {
+		dst.B = append(dst.B, data...)
+		return nil
+	}
+
+	var sw scrapeWork
+	sw.Config = &ScrapeWork{
+		StreamParse: true,
+	}
+	sw.ReadData = readDataFunc
+	sw.PushData = func(_ *auth.Token, _ *prompbmarshal.WriteRequest) {}
```

Comment:
adding artificial delay to PushData with 

`time.Sleep(time.MilliSecond)`
 
 Significantly increases benchmark difference. Proposed changes add x3-x5 better performance.

Reply:
fixed